training/CHAPEL_TUTORIAL.md

```diff
@@ -0,0 +1,861 @@
+
+// you can access an existing array's domain using `.domain`
+var boolArray: [makeDistDom(10)] bool;
+var uintArray: [boolArray.domain] uint;
+```
+Note: `makeDistDom` is a helper function in Arkouda and not part of Chapel.
+
+<a id="init_fact"></a>
+## Initial Factorial
+Let's start our chapel journey by writing a function to calculate the factorial of a given integer `n`.
+Where factorial is
+
+$$ n! = \prod_{i=1}^n i = 1 \cdot 2 \cdot\ldots\cdot (n-1) \cdot n$$
+
+This will introduce the syntax for `proc` and `for` loops!
+
+<a id="ex3"></a>
+#### Example 3: Initial Factorial
+Our initial implementation is similar to other languages:
+```Chapel
+proc factorial(n: int) {
+  var fact: int = 1;
+  for i in 1..n {
+    fact *= i;
+  }
+  return fact;
+}
+writeln(factorial(5));
+```
+```console
+$ chpl Tutorial.chpl -o tutorial
+$ ./tutorial
+120
+```
+
+Excellent! Now let's write this loop in parallel.
+Our parallelism comes from the data parallel [`forall`](https://chapel-lang.org/docs/users-guide/datapar/forall.html) loop.
+The `forall` loop basically handles the parallelism for you!
+
+<a id="ex4"></a>
+#### Example 4: Parallel Factorial Attempt
+So what happens if we just replace  the `for` loop with a `forall`?
+```Chapel
+proc factorial(n: int) {
+  var fact: int = 1;
+  forall i in 1..n {
+    fact *= i;
+  }
+  return fact;
+}
+writeln(factorial(5));
+```
+```console
+$ chpl Tutorial.chpl -o tutorial
+Tutorial.chpl:41: In function 'factorial':
+Tutorial.chpl:44: error: cannot assign to const variable
+Tutorial.chpl:43: note: The shadow variable 'fact' is constant due to forall intents in this loop
+```
+That's not what we want to see! There was an error during compilation.
+The issue is different iterations of the parallel loop could be modifying the outer variable `fact` simultaneously.
+
+Chapel handles this by giving each task its own copy of `fact` known as a [shadow variable](https://chapel-lang.org/docs/primers/forallLoops.html?highlight=intents#task-intents-and-shadow-variables).
+To combine the shadow variables and write the result back into `fact`, we'll need to learn about `reduction` operations.
+
+<a id="reduce_and_scan"></a>
+## `reduce` and `scan`
+[Reductions and scans](https://chapel-lang.org/docs/language/spec/data-parallelism.html?highlight=scan#reductions-and-scans)
+cumulatively apply an operation over the elements of an array (or any iterable) in parallel.
+
+- `op scan array`
+  - scans over `array` and cumulatively applies `op` to every element
+  - returns an array
+  - `+ scan a` behaves like [`np.cumsum(a)`](https://numpy.org/doc/stable/reference/generated/numpy.cumsum.html) in numpy
+- `op reduce array`
+  - reduces the result of a scan to a single summary value
+  - returns a single value
+  - `+ reduce a` behaves like [`sum(a)`](https://docs.python.org/3/library/functions.html#sum) in python
+
+<a id="ex5"></a>
+#### Example 5: `reduce` and `scan`
+```Chapel
+var a: [0..#5] int = [1, 2, 3, 4, 5];
```

Comment:
The `0..#n` syntax is much more common in arkouda, but I do really like `0..<n`. I might start using that enough that it will be needed in the next update of these docs!

Reply:
Arkouda development started very close to the time that `..<` was added to the language, which probably explains its choice.  If the Arkouda devs as a whole agreed with you/us, I think it'd be worth filing a PR to just change all `0..#` occurrences to `0..<`.  Note that `#` is still very useful in other cases like `lo..#segSize`, so I'm not trying to throw the operator under the bus... just not  so wild about it in the `0..n-1` context anymore.